File: backend/services/stigmer-server/pkg/query/activity/handler/handler_test.go
package handler

import (
	"context"
	"fmt"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"google.golang.org/protobuf/types/known/timestamppb"

	activityv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/activity/v1"
	sessionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/session/v1"
	workflowexecutionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/workflowexecution/v1"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource/apiresourcekind"
	"github.com/stigmer/stigmer/backend/libs/go/store"
	"github.com/stigmer/stigmer/backend/libs/go/store/sqlite"
)

// The store-driven tests below exercise the full read path against a real
// SQLite store (the setupTestController precedent): rows are written exactly
// as the persist steps write them, and the handler reads them back through
// store.ListResources — no mocks between the handler and the storage engine.

func setupHandler(t *testing.T) (*Handler, store.Store) {
	t.Helper()
	st, err := sqlite.NewStore(t.TempDir() + "/test.sqlite")
	require.NoError(t, err, "failed to create temp sqlite store")
	t.Cleanup(func() { _ = st.Close() })
	return NewHandler(st), st
}

// auditAt builds the audit shape SetAuditFieldsForCreate stamps, with
// statusAudit.updatedAt at the given instant — the recents sort key.
func auditAt(updated time.Time) *apiresource.ApiResourceAudit {
	stamp := timestamppb.New(updated)
	return &apiresource.ApiResourceAudit{
		SpecAudit:   &apiresource.ApiResourceAuditInfo{CreatedAt: stamp, UpdatedAt: stamp},
		StatusAudit: &apiresource.ApiResourceAuditInfo{CreatedAt: stamp, UpdatedAt: stamp},
	}
}

func saveSession(t *testing.T, st store.Store, id, subject string, labels map[string]string, audit *apiresource.ApiResourceAudit) {
	t.Helper()
	session := &sessionv1.Session{
		ApiVersion: "agentic.stigmer.ai/v1",
		Kind:       "Session",
		Metadata:   &apiresource.ApiResourceMetadata{Id: id, Name: id, Org: "stigmer", Labels: labels},
		Spec:       &sessionv1.SessionSpec{AgentInstanceId: "agi_test", Subject: subject},
		Status:     &apiresource.ApiResourceAuditStatus{Audit: audit},
	}
	require.NoError(t, st.SaveResource(context.Background(), apiresourcekind.ApiResourceKind_session, id, session))
}

func saveExecution(t *testing.T, st store.Store, id, name string, phase workflowexecutionv1.ExecutionPhase, audit *apiresource.ApiResourceAudit) {
	t.Helper()
	execution := &workflowexecutionv1.WorkflowExecution{
		ApiVersion: "agentic.stigmer.ai/v1",
		Kind:       "WorkflowExecution",
		Metadata:   &apiresource.ApiResourceMetadata{Id: id, Name: name, Org: "stigmer"},
		Status:     &workflowexecutionv1.WorkflowExecutionStatus{Phase: phase, Audit: audit},
	}
	require.NoError(t, st.SaveResource(context.Background(), apiresourcekind.ApiResourceKind_workflow_execution, id, execution))
}

func listEntries(t *testing.T, h *Handler, pageSize int32) []*activityv1.RecentActivityEntry {
	t.Helper()
	resp, err := h.ListRecentActivity(context.Background(), &activityv1.ListRecentActivityRequest{PageSize: pageSize})
	require.NoError(t, err)
	return resp.GetEntries()
}

func entryIDs(entries []*activityv1.RecentActivityEntry) []string {
	ids := make([]string, len(entries))
	for i, e := range entries {
		ids[i] = e.GetId()
	}
	return ids
}

func TestListRecentActivity_MergesKindsNewestFirst(t *testing.T) {
	h, st := setupHandler(t)
	base := time.Date(2026, 8, 1, 12, 0, 0, 0, time.UTC)

	saveSession(t, st, "ses_old", "Plan the migration", nil, auditAt(base))
	saveExecution(t, st, "wfx_mid", "nightly-sync", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, auditAt(base.Add(1*time.Hour)))
	saveSession(t, st, "ses_new", "Review the PR", nil, auditAt(base.Add(2*time.Hour)))
	saveExecution(t, st, "wfx_newest", "deploy", workflowexecutionv1.ExecutionPhase_EXECUTION_IN_PROGRESS, auditAt(base.Add(3*time.Hour)))

	entries := listEntries(t, h, 0)

	assert.Equal(t, []string{"wfx_newest", "ses_new", "wfx_mid", "ses_old"}, entryIDs(entries),
		"entries must interleave both kinds sorted by statusAudit.updatedAt descending")
}

func TestListRecentActivity_ProjectionFields(t *testing.T) {
	h, st := setupHandler(t)
	when := time.Date(2026, 8, 1, 12, 0, 0, 0, time.UTC)

	saveSession(t, st, "ses_1", "Plan the migration", nil, auditAt(when))
	saveExecution(t, st, "wfx_1", "nightly-sync", workflowexecutionv1.ExecutionPhase_EXECUTION_FAILED, auditAt(when.Add(time.Hour)))

	entries := listEntries(t, h, 0)
	require.Len(t, entries, 2)

	execution, session := entries[0], entries[1]

	assert.Equal(t, "workflow_execution", execution.GetType())
	assert.Equal(t, "nightly-sync", execution.GetSubject())
	assert.Equal(t, "failed", execution.GetStatus(), "execution status must be the phase display token")
	assert.Equal(t, when.Add(time.Hour).Unix(), execution.GetUpdatedAt().GetSeconds())

	assert.Equal(t, "session", session.GetType())
	assert.Equal(t, "Plan the migration", session.GetSubject())
	assert.Empty(t, session.GetStatus(), "sessions carry no status token")
	assert.Equal(t, when.Unix(), session.GetUpdatedAt().GetSeconds())
}

func TestListRecentActivity_SubjectFallbacks(t *testing.T) {
	h, st := setupHandler(t)
	base := time.Date(2026, 8, 1, 12, 0, 0, 0, time.UTC)

	saveSession(t, st, "ses_empty", "", nil, auditAt(base.Add(3*time.Hour)))
	saveSession(t, st, "ses_sentinel", "Auto-created session", nil, auditAt(base.Add(2*time.Hour)))
	saveExecution(t, st, "wfx_unnamed", "", workflowexecutionv1.ExecutionPhase_EXECUTION_PENDING, auditAt(base.Add(1*time.Hour)))

	entries := listEntries(t, h, 0)
	require.Len(t, entries, 3)

	assert.Equal(t, "Untitled session", entries[0].GetSubject(), "empty subject falls back")
	assert.Equal(t, "Untitled session", entries[1].GetSubject(), "the auto-created sentinel falls back")
	assert.Equal(t, "Untitled execution", entries[2].GetSubject(), "unnamed execution falls back")
}

func TestListRecentActivity_ExcludesRuntimeOriginSessions(t *testing.T) {
	h, st := setupHandler(t)
	when := auditAt(time.Date(2026, 8, 1, 12, 0, 0, 0, time.UTC))

	for i, key := range runtimeOriginLabels {
		saveSession(t, st, fmt.Sprintf("ses_runtime_%d", i), "runtime", map[string]string{key: "some-id"}, when)
	}
	saveSession(t, st, "ses_console", "console", nil, when)
	saveSession(t, st, "ses_labeled", "labeled", map[string]string{"team": "billing"}, when)

	entries := listEntries(t, h, 0)

	assert.ElementsMatch(t, []string{"ses_console", "ses_labeled"}, entryIDs(entries),
		"every runtime-origin label must exclude its session; unrelated labels must not")
}

func TestListRecentActivity_PageSize(t *testing.T) {
	h, st := setupHandler(t)
	base := time.Date(2026, 8, 1, 0, 0, 0, 0, time.UTC)

	// 35 sessions, newest last — enough to cross the default page size of 30.
	for i := 0; i < 35; i++ {
		saveSession(t, st, fmt.Sprintf("ses_%02d", i), "s", nil, auditAt(base.Add(time.Duration(i)*time.Minute)))
	}

	t.Run("explicit_page_size_trims", func(t *testing.T) {
		entries := listEntries(t, h, 2)
		assert.Equal(t, []string{"ses_34", "ses_33"}, entryIDs(entries))
	})

	t.Run("zero_defaults_to_30", func(t *testing.T) {
		entries := listEntries(t, h, 0)
		require.Len(t, entries, 30)
		assert.Equal(t, "ses_34", entries[0].GetId(), "page must contain the newest rows")
		assert.Equal(t, "ses_05", entries[29].GetId())
	})

	t.Run("negative_defaults_to_30", func(t *testing.T) {
		assert.Len(t, listEntries(t, h, -5), 30)
	})

	t.Run("oversized_request_returns_all_rows_under_cap", func(t *testing.T) {
		assert.Len(t, listEntries(t, h, 500), 35, "requests above the cap clamp to 100, which exceeds the 35 stored rows")
	})
}

func TestListRecentActivity_TimestampFallbackAndTies(t *testing.T) {
	h, st := setupHandler(t)
	base := time.Date(2026, 8, 1, 12, 0, 0, 0, time.UTC)

	t.Run("falls_back_to_spec_created_at", func(t *testing.T) {
		// A row whose statusAudit was never stamped (external tooling / older
		// build) sorts by specAudit.createdAt instead of sinking to epoch.
		saveSession(t, st, "ses_no_status_audit", "legacy", nil, &apiresource.ApiResourceAudit{
			SpecAudit: &apiresource.ApiResourceAuditInfo{CreatedAt: timestamppb.New(base.Add(2 * time.Hour))},
		})
		saveSession(t, st, "ses_stamped", "current", nil, auditAt(base.Add(1*time.Hour)))

		entries := listEntries(t, h, 0)
		assert.Equal(t, []string{"ses_no_status_audit", "ses_stamped"}, entryIDs(entries))
	})

	t.Run("equal_timestamps_keep_sessions_before_executions", func(t *testing.T) {
		tie := auditAt(base.Add(3 * time.Hour))
		saveSession(t, st, "ses_tie", "tie", nil, tie)
		saveExecution(t, st, "wfx_tie", "tie", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, tie)

		entries := listEntries(t, h, 2)
		assert.Equal(t, []string{"ses_tie", "wfx_tie"}, entryIDs(entries),
			"the sort is stable over the sessions-then-executions load order (cloud tie-break parity)")
	})
}

func TestListRecentActivity_EmptyStore(t *testing.T) {
	h, _ := setupHandler(t)
	entries := listEntries(t, h, 0)
	assert.Empty(t, entries, "an empty store answers an empty page, not an error")
}

func TestNormalizePageSize(t *testing.T) {
	tests := []struct {
		name      string
		requested int32
		want      int
	}{
		{"zero_defaults", 0, defaultPageSize},
		{"negative_defaults", -1, defaultPageSize},
		{"in_range_passes_through", 7, 7},
		{"at_cap_passes_through", 100, 100},
		{"above_cap_clamps", 101, maxPageSize},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			assert.Equal(t, tt.want, normalizePageSize(tt.requested))
		})
	}
}

func TestResolvePhase(t *testing.T) {
	tests := []struct {
		phase workflowexecutionv1.ExecutionPhase
		want  string
	}{
		{workflowexecutionv1.ExecutionPhase_EXECUTION_PENDING, "pending"},
		{workflowexecutionv1.ExecutionPhase_EXECUTION_IN_PROGRESS, "running"},
		{workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "completed"},
		{workflowexecutionv1.ExecutionPhase_EXECUTION_FAILED, "failed"},
		{workflowexecutionv1.ExecutionPhase_EXECUTION_CANCELLED, "cancelled"},
		{workflowexecutionv1.ExecutionPhase_EXECUTION_TERMINATED, "terminated"},
		{workflowexecutionv1.ExecutionPhase_EXECUTION_PAUSED, "paused"},
		{workflowexecutionv1.ExecutionPhase_EXECUTION_PHASE_UNSPECIFIED, "unknown"},
		{workflowexecutionv1.ExecutionPhase(99), "unknown"},
	}
	for _, tt := range tests {
		t.Run(tt.want, func(t *testing.T) {
			assert.Equal(t, tt.want, resolvePhase(tt.phase))
		})
	}
}

func TestTimestampAfter(t *testing.T) {
	tests := []struct {
		name string
		a, b *timestamppb.Timestamp
		want bool
	}{
		{"later_seconds", &timestamppb.Timestamp{Seconds: 2}, &timestamppb.Timestamp{Seconds: 1}, true},
		{"earlier_seconds", &timestamppb.Timestamp{Seconds: 1}, &timestamppb.Timestamp{Seconds: 2}, false},
		{"same_seconds_later_nanos", &timestamppb.Timestamp{Seconds: 1, Nanos: 2}, &timestamppb.Timestamp{Seconds: 1, Nanos: 1}, true},
		{"equal_is_not_after", &timestamppb.Timestamp{Seconds: 1, Nanos: 1}, &timestamppb.Timestamp{Seconds: 1, Nanos: 1}, false},
		{"nil_is_zero", nil, &timestamppb.Timestamp{Seconds: 1}, false},
		{"after_nil", &timestamppb.Timestamp{Seconds: 1}, nil, true},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			assert.Equal(t, tt.want, timestampAfter(tt.a, tt.b))
		})
	}
}

File: backend/services/stigmer-server/src/query/activity/__tests__/handler.test.ts
/**
 * Pins the activity handler against Go's
 * pkg/query/activity/handler/handler_test.go, case-for-case: the two-kind
 * newest-first merge, the projection fields, the subject sentinels, the
 * runtime-origin exclusions, the page-size table (default 30 / cap 100 —
 * the constants conformance deliberately does not drive over the wire),
 * the timestamp fallback + tie ordering (sessions before executions), the
 * empty store, and the resolvePhase / timestampAfter tables.
 */
import { create } from "@bufbuild/protobuf";
import { TimestampSchema } from "@bufbuild/protobuf/wkt";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import { ListRecentActivityRequestSchema } from "@stigmer/protos/ai/stigmer/activity/v1/io_pb";
import { SessionSchema } from "@stigmer/protos/ai/stigmer/agentic/session/v1/api_pb";
import { WorkflowExecutionSchema } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/api_pb";
import { ExecutionPhase } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/enum_pb";
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";

import { createLogger } from "../../../boot/logger.js";
import {
  tempStore,
  type TempStore,
} from "../../../store/sqlite/__tests__/support.js";
import {
  ActivityHandler,
  DEFAULT_PAGE_SIZE,
  MAX_PAGE_SIZE,
  normalizePageSize,
  resolvePhase,
  resolveSubject,
  timestampAfter,
} from "../handler.js";

const silentLogger = createLogger({
  level: "error",
  pretty: false,
  write: () => undefined,
});

function request(pageSize: number) {
  return create(ListRecentActivityRequestSchema, { pageSize, org: "" });
}

let temp: TempStore;
let handler: ActivityHandler;

beforeEach(() => {
  temp = tempStore();
  handler = new ActivityHandler(temp.store, silentLogger);
});

afterEach(async () => {
  await temp.cleanup();
});

async function seedSession(
  id: string,
  opts?: {
    subject?: string;
    labels?: Record<string, string>;
    statusUpdatedAtSeconds?: number;
    statusUpdatedAtNanos?: number;
    specCreatedAtSeconds?: number;
  },
): Promise<void> {
  const statusAudit =
    opts?.statusUpdatedAtSeconds === undefined
      ? undefined
      : {
          updatedAt: {
            seconds: BigInt(opts.statusUpdatedAtSeconds),
            nanos: opts.statusUpdatedAtNanos ?? 0,
          },
        };
  const specAudit =
    opts?.specCreatedAtSeconds === undefined
      ? undefined
      : { createdAt: { seconds: BigInt(opts.specCreatedAtSeconds) } };
  const session = create(SessionSchema, {
    metadata: { id, name: id, org: "acme", labels: opts?.labels ?? {} },
    spec: { subject: opts?.subject ?? "", agentInstanceId: "agi_test" },
    status: { audit: { statusAudit, specAudit } },
  });
  await temp.store.saveResource(
    ApiResourceKind.session,
    id,
    SessionSchema,
    session,
  );
}

async function seedExecution(
  id: string,
  opts?: {
    name?: string;
    phase?: ExecutionPhase;
    statusUpdatedAtSeconds?: number;
  },
): Promise<void> {
  const execution = create(WorkflowExecutionSchema, {
    metadata: { id, name: opts?.name ?? id, org: "acme" },
    status: {
      phase: opts?.phase ?? ExecutionPhase.EXECUTION_COMPLETED,
      audit:
        opts?.statusUpdatedAtSeconds === undefined
          ? undefined
          : {
              statusAudit: {
                updatedAt: { seconds: BigInt(opts.statusUpdatedAtSeconds) },
              },
            },
    },
  });
  await temp.store.saveResource(
    ApiResourceKind.workflow_execution,
    id,
    WorkflowExecutionSchema,
    execution,
  );
}

describe("listRecentActivity (Go handler_test.go)", () => {
  it("merges both kinds newest-first", async () => {
    await seedSession("ses_old", {
      subject: "oldest",
      statusUpdatedAtSeconds: 100,
    });
    await seedExecution("wfe_mid", { statusUpdatedAtSeconds: 200 });
    await seedSession("ses_new", {
      subject: "newest",
      statusUpdatedAtSeconds: 300,
    });

    const response = await handler.listRecentActivity(request(100));

    expect(response.entries.map((entry) => entry.id)).toEqual([
      "ses_new",
      "wfe_mid",
      "ses_old",
    ]);
  });

  it("projects the sidebar fields per kind", async () => {
    await seedSession("ses_1", {
      subject: "Plan the migration",
      statusUpdatedAtSeconds: 100,
    });
    await seedExecution("wfe_1", {
      name: "nightly-sync",
      phase: ExecutionPhase.EXECUTION_IN_PROGRESS,
      statusUpdatedAtSeconds: 200,
    });

    const response = await handler.listRecentActivity(request(100));

    const session = response.entries.find((entry) => entry.id === "ses_1");
    expect(session?.type).toBe("session");
    expect(session?.subject).toBe("Plan the migration");
    expect(session?.status).toBe("");
    expect(session?.updatedAt).toBeDefined();

    const execution = response.entries.find((entry) => entry.id === "wfe_1");
    expect(execution?.type).toBe("workflow_execution");
    expect(execution?.subject).toBe("nightly-sync");
    expect(execution?.status).toBe("running");
  });

  it("maps the subject sentinels to display placeholders", async () => {
    await seedSession("ses_auto", {
      subject: "Auto-created session",
      statusUpdatedAtSeconds: 100,
    });
    await seedSession("ses_empty", {
      subject: "",
      statusUpdatedAtSeconds: 200,
    });
    await seedExecution("wfe_unnamed", {
      name: "",
      statusUpdatedAtSeconds: 300,
    });

    const response = await handler.listRecentActivity(request(100));
    const byId = new Map(response.entries.map((entry) => [entry.id, entry]));

    expect(byId.get("ses_auto")?.subject).toBe("Untitled session");
    expect(byId.get("ses_empty")?.subject).toBe("Untitled session");
    expect(byId.get("wfe_unnamed")?.subject).toBe("Untitled execution");
  });

  it("excludes runtime-origin sessions for every label key", async () => {
    await seedSession("ses_console", {
      subject: "personal",
      statusUpdatedAtSeconds: 100,
    });
    const labeled: Array<[id: string, key: string]> = [
      ["ses_channel", "stigmer.ai/channel-id"],
      ["ses_share", "stigmer.ai/share-id"],
      ["ses_guest", "stigmer.ai/guest-cookie-id"],
      ["ses_schedule", "stigmer.ai/schedule-id"],
    ];
    for (const [id, key] of labeled) {
      await seedSession(id, {
        subject: "runtime",
        labels: { [key]: "x" },
        statusUpdatedAtSeconds: 200,
      });
    }

    const response = await handler.listRecentActivity(request(100));
    const ids = response.entries.map((entry) => entry.id);

    expect(ids).toContain("ses_console");
    for (const [id] of labeled) {
      expect(ids, `${id} must be excluded`).not.toContain(id);
    }
  });

  it("trims to page_size keeping the newest; zero and negatives mean the default page", async () => {
    for (let i = 0; i < 3; i++) {
      await seedSession(`ses_${i}`, {
        subject: `s${i}`,
        statusUpdatedAtSeconds: 100 + i,
      });
    }

    const trimmed = await handler.listRecentActivity(request(1));
    expect(trimmed.entries.map((entry) => entry.id)).toEqual(["ses_2"]);

    for (const pageSize of [0, -5]) {
      const response = await handler.listRecentActivity(request(pageSize));
      expect(response.entries).toHaveLength(3);
    }

    const oversize = await handler.listRecentActivity(request(100_000));
    expect(oversize.entries).toHaveLength(3);
  });

  it("falls back to specAudit.createdAt when the status audit is unstamped", async () => {
    await seedSession("ses_stamped", {
      subject: "stamped",
      statusUpdatedAtSeconds: 100,
    });
    await seedSession("ses_fallback", {
      subject: "fallback",
      specCreatedAtSeconds: 200,
    });

    const response = await handler.listRecentActivity(request(100));

    expect(response.entries.map((entry) => entry.id)).toEqual([
      "ses_fallback",
      "ses_stamped",
    ]);
  });

  it("keeps sessions before executions on equal timestamps (stable sort)", async () => {
    await seedExecution("wfe_tie", { statusUpdatedAtSeconds: 100 });
    await seedSession("ses_tie", {
      subject: "tie",
      statusUpdatedAtSeconds: 100,
    });

    const response = await handler.listRecentActivity(request(100));

    // Sessions load first (Go's append order), and the stable sort keeps
    // insertion order on ties.
    expect(response.entries.map((entry) => entry.id)).toEqual([
      "ses_tie",
      "wfe_tie",
    ]);
  });

  it("answers an empty store with an empty page", async () => {
    const response = await handler.listRecentActivity(request(10));
    expect(response.entries).toEqual([]);
  });
});

describe("normalizePageSize (Go's table)", () => {
  it("folds out-of-range values into range", () => {
    expect(normalizePageSize(0)).toBe(DEFAULT_PAGE_SIZE);
    expect(normalizePageSize(-1)).toBe(DEFAULT_PAGE_SIZE);
    expect(normalizePageSize(1)).toBe(1);
    expect(normalizePageSize(30)).toBe(30);
    expect(normalizePageSize(100)).toBe(MAX_PAGE_SIZE);
    expect(normalizePageSize(101)).toBe(MAX_PAGE_SIZE);
    expect(normalizePageSize(100_000)).toBe(MAX_PAGE_SIZE);
  });
});

describe("resolveSubject", () => {
  it("maps the sentinel and emptiness to the placeholder, passes real titles", () => {
    expect(resolveSubject("")).toBe("Untitled session");
    expect(resolveSubject("Auto-created session")).toBe("Untitled session");
    expect(resolveSubject("Real title")).toBe("Real title");
  });
});

describe("resolvePhase (Go's table)", () => {
  it("maps every phase to its badge token, unknowns to 'unknown'", () => {
    expect(resolvePhase(ExecutionPhase.EXECUTION_PENDING)).toBe("pending");
    expect(resolvePhase(ExecutionPhase.EXECUTION_IN_PROGRESS)).toBe("running");
    expect(resolvePhase(ExecutionPhase.EXECUTION_COMPLETED)).toBe("completed");
    expect(resolvePhase(ExecutionPhase.EXECUTION_FAILED)).toBe("failed");
    expect(resolvePhase(ExecutionPhase.EXECUTION_CANCELLED)).toBe("cancelled");
    expect(resolvePhase(ExecutionPhase.EXECUTION_TERMINATED)).toBe(
      "terminated",
    );
    expect(resolvePhase(ExecutionPhase.EXECUTION_PAUSED)).toBe("paused");
    expect(resolvePhase(ExecutionPhase.EXECUTION_PHASE_UNSPECIFIED)).toBe(
      "unknown",
    );
    expect(resolvePhase(999 as ExecutionPhase)).toBe("unknown");
  });
});

describe("timestampAfter (Go's table)", () => {
  const ts = (seconds: number, nanos = 0) =>
    create(TimestampSchema, { seconds: BigInt(seconds), nanos });

  it("compares seconds first, then nanos, strictly", () => {
    expect(timestampAfter(ts(2), ts(1))).toBe(true);
    expect(timestampAfter(ts(1), ts(2))).toBe(false);
    expect(timestampAfter(ts(1, 5), ts(1, 3))).toBe(true);
    expect(timestampAfter(ts(1, 3), ts(1, 5))).toBe(false);
    // Strictly after: equality is false on both sides (the stable-sort
    // tie contract depends on it).
    expect(timestampAfter(ts(1, 1), ts(1, 1))).toBe(false);
    expect(timestampAfter(undefined, undefined)).toBe(false);
  });
});
